web-api/src/mind/state.rs

```diff
@@ -0,0 +1,197 @@
+            storage::Document::get_personalized(&self.storage, embeddings.iter().map(|(id, _)| id))
+                .await?
+                .into_iter()
+                .map(|document| (document.id.clone(), document))
+                .collect::<HashMap<_, _>>();
+        let documents = embeddings
+            .into_iter()
+            .map(|(id, embedding)| {
+                let document = documents.remove(&id).unwrap(/* document must already exist */);
+                let document = IngestedDocument {
+                    id,
+                    snippet: String::new(/* unused for in-memory db */),
+                    properties: document.properties,
+                    tags: document.tags,
+                };
+                (document, embedding)
+            })
+            .collect_vec();
+        storage::Document::insert(&self.storage, documents).await?;
+
+        Ok(())
+    }
+
+    pub(super) async fn interact(
+        &self,
+        user: &UserId,
+        documents: impl IntoIterator<Item = (&DocumentId, DateTime<Utc>)>,
+    ) -> Result<(), Panic> {
+        for (id, time) in documents {
+            update_interactions(
+                &self.storage,
+                &self.coi,
+                user,
+                [&(id.clone(), UserInteractionType::Positive)],
+                self.personalization.store_user_history,
+                time,
+            )
+            .await?;
+        }
+
+        Ok(())
+    }
+
+    pub(super) async fn personalize(
+        &self,
+        user: &UserId,
+        by: PersonalizeBy<'_>,
+        time: DateTime<Utc>,
+    ) -> Result<Option<Vec<DocumentId>>, Panic> {
+        personalize_documents_by(
+            &self.storage,
+            &self.coi,
+            user,
+            &self.personalization,
+            by,
+            time,
+        )
+        .await
+        .map(|documents| {
+            documents.map(|documents| documents.into_iter().map(|document| document.id).collect())
+        })
+        .map_err(Into::into)
+    }
+}
+
+/// The results of iteration of the saturation benchmark.
+#[derive(Debug, Default, Serialize)]
+pub(super) struct SaturationIteration {
+    pub(super) shown_documents: Vec<DocumentId>,
+    pub(super) clicked_documents: Vec<DocumentId>,
+}
+
+/// The results of the saturation benchmark for one topic.
+#[derive(Debug, Default, Deref, DerefMut, Serialize)]
+pub(super) struct SaturationTopicResult {
+    topic: String,
+    #[deref]
+    #[deref_mut]
+    iterations: Vec<SaturationIteration>,
+}
+
+impl SaturationTopicResult {
+    pub(super) fn new(topic: &str, iterations: usize) -> Self {
+        Self {
+            topic: topic.to_string(),
+            iterations: Vec::with_capacity(iterations),
+        }
+    }
+}
+
+/// The results of the saturation benchmark.
+#[derive(Debug, Default, Deref, DerefMut, Serialize)]
+pub(super) struct SaturationResult {
+    topics: Vec<SaturationTopicResult>,
+}
+
+impl SaturationResult {
```

Comment:
We could move the result structs to `data.rs` as they hold data in some way but don't have a strong opinion about it

Reply:
i guess so, i neither have a strong opinion on that. i placed it there since it holds the state of the iteration outputs, but it could as well be moved into `data.rs`.